lib/events/setter.go

```diff
@@ -0,0 +1,148 @@
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+func NewSetter(cfg SetterConfig) (*Setter, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &Setter{
+		cfg: cfg,
+	}, nil
+}
+
+// SetterConfig configures an event setter
+type SetterConfig struct {
+	// SessionID defines the session to record.
+	SessionID session.ID
+
+	// ServerID is a server ID to write
+	ServerID string
+
+	// Namespace is the session namespace.
+	Namespace string
+
+	// Clock is used to override time in tests
+	Clock clockwork.Clock
+
+	// UID is UID generator
+	UID utils.UID
+
+	// ClusterName defines the name of this teleport cluster.
+	ClusterName string
+}
+
+// CheckAndSetDefaults checks and sets defaults
+func (cfg *SetterConfig) CheckAndSetDefaults() error {
+	if cfg.SessionID.IsZero() {
+		return trace.BadParameter("event checker config: missing parameter SessionID")
+	}
+	if cfg.ClusterName == "" {
+		return trace.BadParameter("event checker config: missing parameter ClusterName")
+	}
+	if cfg.Namespace == "" {
+		cfg.Namespace = apidefaults.Namespace
+	}
+	if cfg.Clock == nil {
+		cfg.Clock = clockwork.NewRealClock()
+	}
+	if cfg.UID == nil {
+		cfg.UID = utils.NewRealUID()
+	}
+
+	return nil
+}
+
+type Setter struct {
+	mtx            sync.Mutex
+	cfg            SetterConfig
+	lastPrintEvent *apievents.SessionPrint
+	eventIndex     atomic.Uint64
+}
+
+// SetupSessionEvent will set necessary event fields for session-related
+// events and must be called before the event is recorded, regardless
+// of whether the event will be recorded, emitted, or both.
+func (c *Setter) SetupSessionEvent(event apievents.AuditEvent) (apievents.AuditEvent, error) {
+	if err := checkAndSetEventFields(event, c.cfg.Clock, c.cfg.UID, c.cfg.ClusterName); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	sess, ok := event.(SessionMetadataSetter)
+	if ok {
+		sess.SetSessionID(string(c.cfg.SessionID))
+	}
+
+	srv, ok := event.(ServerMetadataSetter)
+	if ok {
+		srv.SetServerNamespace(c.cfg.Namespace)
+		srv.SetServerID(c.cfg.ServerID)
+	}
+
+	// ensure index is incremented and loaded atomically
+	event.SetIndex(int64(c.eventIndex.Add(1) - 1))
+
+	printEvent, ok := event.(*apievents.SessionPrint)
+	if !ok {
+		return event, nil
+	}
+
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	if c.lastPrintEvent != nil {
+		printEvent.Offset = c.lastPrintEvent.Offset + int64(len(c.lastPrintEvent.Data))
+		printEvent.DelayMilliseconds = diff(c.lastPrintEvent.Time, printEvent.Time) + c.lastPrintEvent.DelayMilliseconds
+		printEvent.ChunkIndex = c.lastPrintEvent.ChunkIndex + 1
+	}
+	c.lastPrintEvent = printEvent
+
+	return event, nil
+}
+
+type SetterAndRecorder struct {
+	SessionEventSetter
+	SessionRecorder
+}
+
+// SetupAndRecordEvent will set necessary event fields for session-related
+// events and record them.
+func (s SetterAndRecorder) SetupAndRecordEvent(ctx context.Context, event apievents.AuditEvent) error {
```

Comment:
So if you removed `SetupAndRecordEvent` from `SessionSetterRecorder`, then this could just be a free function:

```
func SetupAndRecord(ctx context.Context, s SessionSetterRecorder, event apievents.AuditEvent) error {
    // the body remains the same, but now you no longer
    // need a new struct for SetterAndRecorder
}
```

Reply:
That's a good idea, done